### ISSUE TITLE
ActiveRecord: calling pluck in class method duplicates limit and offset bindings, resulting in invalid database queries

### ISSUE THREAD

**Issue by grk** [rails/rails#29741]
### Steps to reproduce
Run https://gist.github.com/grk/85714511f79e5e7f40a7f05109966bb0

### Expected behavior
Should run without problems just as the same query not ran through a class method:
```
(0.1ms)  SELECT  "posts"."id" FROM "posts" INNER JOIN "posts" "children_posts" ON "children_posts"."parent_id" = "posts"."id" WHERE "posts"."parent_id" IS NULL LIMIT ? OFFSET ?  [["LIMIT", 20], ["OFFSET", 0]]
```

### Actual behavior
Crashes with
```
BugTest#test_association_stuff:
ActiveRecord::StatementInvalid: SQLite3::RangeException: bind or column index out of range: SELECT  "posts"."id" FROM "posts" INNER JOIN "posts" "children_posts" ON "children_posts"."parent_id" = "posts"."id" AND "posts"."parent_id" IS NULL WHERE "posts"."parent_id" IS NULL LIMIT ? OFFSET ?
```

### System configuration
**Rails version**: 5.1.2

**Ruby version**: 2.4.1

**Comment by EddM** [comment]
Just spent a couple of hours tracking this down. No solution yet, but some notes that may or may not help someone or myself at a later time:

- This doesn't have anything to do with `#pluck` in particular, or even `#all`. If you swap out the `pluck_id` method in the example with `where("").to_a`, for example, same thing.
- Closest I can get is `ActiveRecord::Relation#exec_queries` (activerecord/lib/active_record/relation.rb:659)
  - `#bound_attributes` (`ActiveRecord::QueryMethods.bound_attributes`) is what's returning the duplicated `OFFSET` and `LIMIT` binds
    - Line 94, `join_clause: arel.bind_values` is returning these extra offset/limit clauses. The naming of that hash key (`join_clause:`) suggests this is erroneous, and that only a `JOIN` clause is expected to be coming out of `arel.bind_values`?

To confirm, this does only happen when you're doing it inside a class method on the model, not when you're running (what should be) an equivalent query outside of it.

This is probably something to do with how `method_missing` is handled in `activerecord/lib/active_record/relation/delegation.rb:87`. I wish I was smart enough to figure out what it is - I'll take a further look tomorrow.

**Comment by grk** [comment]
It's definately an issue with `scoping`, this also fails:

```ruby
    Post.where(parent_id: nil).joins(:children).limit(20).offset(0).scoping do
      Post.pluck(:id)
    end
```

**Comment by grk** [comment]
Further narrowing it down:

```
[1] pry(#<BugTest>)> Post.where(parent_id: nil).joins(:children).limit(20).offset(0).inspect
D, [2017-07-11T11:43:44.373982 #54492] DEBUG -- :   Post Load (0.3ms)  SELECT  "posts".* FROM "posts" INNER JOIN "posts" "children_posts" ON "children_posts"."parent_id" = "posts"."id" WHERE "posts"."parent_id" IS NULL LIMIT ? OFFSET ?  [["LIMIT", 11], ["OFFSET", 0]]
=> "#<ActiveRecord::Relation []>"
[2] pry(#<BugTest>)> Post.current_scope = Post.where(parent_id: nil).joins(:children).limit(20).offset(0)
D, [2017-07-11T11:44:00.040127 #54492] DEBUG -- :   Post Load (0.3ms)  SELECT  "posts".* FROM "posts" INNER JOIN "posts" "children_posts" ON "children_posts"."parent_id" = "posts"."id" AND "posts"."parent_id" IS NULL WHERE "posts"."parent_id" IS NULL LIMIT ? OFFSET ?  [["LIMIT", 20], ["OFFSET", 0], ["LIMIT", 20], ["OFFSET", 0]]
=> #<Post::ActiveRecord_Relation:0x3fe8a6f5c05c>
```

**Comment by che-burashco** [comment]
Hey guys, I think I have a pull request that should fix the problem. It's basically the current_scope as @grk has mentioned, it has the limit and offset attached to it, since it's the same klass that we join to, which leads to a duplication of the limit and offset binds in join_constraints

**Comment by sgrif** [comment]
Fixed by 5c71000d086cc42516934415b79380c2224e1614
